### ISSUE TITLE
[filter framework] hiding all caches is not possible 

### ISSUE THREAD

**Issue by fm-sys** [cgeo/cgeo#10963]
We had several issues where users want to easily hide all caches. I would have expected that this is possible by deselecting all cache types, but if I apply the filter with nothing selected, all caches are still shown at the live map... 

![Screenshot_20210619_194214_cgeo.geocaching_copy_270x585.jpg](https://user-images.githubusercontent.com/64581222/122651495-8040bf00-d139-11eb-9342-192fa2200e0f.jpg)

**Comment by eddiemuc** [comment]
True, "none selected" is currently treated like "all selected" because I couldn't think of any scenario in which someone wants to select "nothing" and this keep the result empty.

Same concept is used for all filters where a "multiple choice" is applied, e.g. sizes

**Comment by fm-sys** [comment]
> couldn't think of any scenario in which someone wants to select "nothing"

Many users use c:geo for planning hiking/mountainbike tours independent of geocaches. Furthermore, a totally cluttered map can be quite annoying when trying to find a specific place at the map. 

I remember a quote from you: "Don't block functionality, just because we can't think of a use case from our point of view" (or something like that) ;-)

**Comment by moving-bits** [comment]
IMHO that's more of a function that should go into the "map quick settings" popup, not into the filters.
I cannot think of any use for it outside the map, but there it would be handy.

**Comment by fm-sys** [comment]
> I cannot think of any use for it outside the map, but there it would be handy.

That's exactly the point. We often over-complicate things because we we try to optimize edge cases with not understandable exceptions. 

IMHO it's a clear thing: If the user explicitly choose to show no cache types there is absolutely no reason to do things differently than the user has decided. 

Place such a filter as quick settings entry would be rather random. a) such a setting would not be used enough for being worth to take away the rare space in the popup and b) there are many other filters which have at least the same priority and would need to go into the quick settings as well...

**Comment by eddiemuc** [comment]
Well, it can be added to tread "nothing-is-selected" as "filter-out-everything" of course. Assuming that this logic should not be applied to all filters (e.g. "Origin" or "contained in stored list" comes to mind) this would be however an added complexity at least in the code. But not too much :-)

**Comment by fm-sys** [comment]
> Assuming that this logic should not be applied to all filters (e.g. "Origin" or "contained in stored list" comes to mind)

Yes, for such filters blank/default does IMHO mean don't filter out anything (show all). The difference between these filters and cachetype filter is, that for the type filter the checkboxes are always there and activated by default seemly providing the possibility to hide everything... 

> Same concept is used for all filters where a "multiple choice" is applied, e.g. sizes

So to be consistent, same concept might apply to size and the status enabled-disabled-archived toogle group, but definitely not the other filters... 

BTW, my suggested behavior is already somehow implemented for "the status enabled-disabled-archived toogle group", just for type and size it is currently inconsistent.

**Comment by Lineflyer** [comment]
> IMHO that's more of a function that should go into the "map quick settings" popup, not into the filters.
> I cannot think of any use for it outside the map, but there it would be handy.

That is actually a much better idea than requiring to construct a filter with no results. I like that.
Not sure though whether it would be more suitable in the map quick settings or just in the map menu (comparable to "disable live"/"enable live").

**Comment by fm-sys** [comment]
See my post above. A top level menu entry is probably not worth it. And even if, it should be additional. 

Selecting no cache types should result in no caches at the live map. That's the only thing which would be consistent...

**Comment by Lineflyer** [comment]
> See my post above. A top level menu entry is probably not worth it. And even if, it should be additional.
> Selecting no cache types should result in no caches at the live map. That's the only thing which would be consistent...

I did not stumble over this exact issue and in contrast to my first posting I have to admit, that the filter result is not logical.
So rather independently of whether it deserves a quick setting, I think the filter should also work in case of "no cache type selected". That would be a sufficient solution for that edge use case.

**Comment by Lineflyer** [comment]
BTW: Same problem was just mentioned on Facebook by a user which does also see this as a bug.

**Comment by eddiemuc** [comment]
Summarizing the situation of this issue:
* As of now, for **all** filters where the user can select from a range of values and no value is selected, this is treated as "apply no filter". In the case where all selectable values cover "all caches" this is identical to the behaviour when all choices are selected. This is not only true for cache type filter, but also e.g. for cache size, origin and stored list filter
* This item wants to change that: the behaviour of "no item selected" should be "apply a filter which always delivers false"

While thinking about #11347 I realized that I don't currently know how to apply this behaviour to a filter where not all selectable items are visible. Let's take "stored list filter" as an example. I have three questions.

Currently, the "stored list filter" considers non-visible item to be "unselected". Thus, when you add such a filter (and no list is initially visible) and just apply the filter, this currently means "apply no filter". In the future, this would mean "filter all caches".
Is this the wanted behaviour?

![image](https://user-images.githubusercontent.com/6909759/127738019-88e218fb-a7cf-4fc2-9f09-3b8629f353aa.png)

Now I use the "add item" to add some of my lists, but I deselect them. Currently, this would also lead to "apply no filter". In the future this would mean "filter all caches". Is this the wanted behaviour?

![image](https://user-images.githubusercontent.com/6909759/127738048-f858a686-6061-433c-a773-12a5b09d58b8.png)

If I add ALL my lists to the screen, the "add item" button vanishes of course. Then the GUI will be "logically identical" to cache type filter (although in one column, not two). But in contrast to "cache type" there are caches NOT in any list at all (e.g. those which are not stored on device). If NO item is selected, should this then filter ALL caches or only those stored in ANY list?

![image](https://user-images.githubusercontent.com/6909759/127738113-ee507e24-cbe2-4103-abce-5733d3db21dd.png)

My personal conclusion was up to now: if I simply tread the "none item selected" as "don't filter", then all the questions above are very easy to answer. As you know I like things simple :-). I am in fear that when we don't want to apply this logic any more, things get complicated (both for user to understand and for implementation). Looking forward to your concepts.

**Comment by eddiemuc** [comment]
And while wer're about to discuss this, I also want to bring range filters into this. For example, the D/T filter will currently apply NO filter if the whole range is selected. E.g. the following setting will not filter at all:

![image](https://user-images.githubusercontent.com/6909759/127738295-733d2b32-875c-433c-b705-816e0ef34eb1.png)

This specifically means that also caches with NO D/T value (e.g. user-defined caches) or the few caches with a D/T value > 5 (e.g. from EC) will be contained in the result.

Filtering only starts when not the full range is selected (e.g. min >1.0 or max <5.0).

**Comment by eddiemuc** [comment]
Two more things: 
* I want to point out that if you currently deselect all items on the cache type filter, then save/apply filter and then go back to the filter view -> then all cache types are selected again. This was implemented by me on purpose (to "underline" that "select none" means "apply no filter")
* The filters on geocaching.com and oc.de "solve" this problem in another way: they simply don't allow selection of "no cache type" (filter can't be applied then). I personally don't like this solution very much, it does not feel user-friendly to me, but maybe this is an acceptable way for the community?

**Comment by eddiemuc** [comment]
And another thing (sorry, my thoughts are coming in quite randomized): 

Coming back to the original request "easily hide all caches":

First, this is very easy to achieve in current setup: just use e.g. cache type filter with all/none items selected and choose "invert filter". Voila, all caches gone

Second, and more important: the only thing I can think of where a user wants to hide all caches is when he wants to use c:geo as a pure map application (e.g. going on a cache-free hiking trip). But using the filtering system to get rid of displayed caches is a really bad idea - because e.g. in live map c:geo will still issue requests to the servers (using up bandwidth and data plan) although this would be totally unnecessary. If c:geo should really provide a "cache-free" mode I would say this is a totally valid request - but the filter system should not be abused for that.

**Comment by Lineflyer** [comment]
> If c:geo should really provide a "cache-free" mode I would say this is a totally valid request - but the filter system should not be abused for that.

Fair enough.
Most of the time I hear this requests from our users its more like "I want to get an overview of my surroundings on the map but the icons clutter the map such, that I can't read it."
So I guess the need is more like a very time-limited hiding possibility: Easily hide them, scroll/look around on the map, once you found the point you want to look at, easily switch it on."

**Comment by moving-bits** [comment]
> Most of the time I hear this requests from our users its more like "I want to get an overview of my surroundings on the map but the icons clutter the map such, that I can't read it."
>
> So I guess the need is more like a very time-limited hiding possibility: Easily hide them, scroll/look around on the map, once you found the point you want to look at, easily switch it on."

That's exactly my use-case as well. Disable any caches, search for the city/... I'm looking for, and then re-enable caches. And at least for me it's not an edge-case, but I use it pretty frequently.

> This can for sure be done outside the filter and would even be more easy than constructing a filter which gives no result.

Agreed. That's why I suggested to put it into map quick settings initially:

> IMHO that's more of a function that should go into the "map quick settings" popup, not into the filters.
> I cannot think of any use for it outside the map, but there it would be handy.

I'm using the quick settings quite frequently for both planning a cache tour and during a tour.

We had discussed renaming the popup's entries from "show xxx" to "hide xxx" already a few of times.
If doing so, a "hide all" option would fit in neatly.

**Comment by fm-sys** [comment]
> BTW: Same problem was just mentioned on Facebook by a user which does also see this as a bug.

Users should IMHO always have the highest priority. If users think something is a bug, it simply _is_ a bug or is at least not understandable enough, even if it was designed that way.

> As of now, for all filters where the user can select from a range of values and no value is selected, this is treated as "apply no filter". In the case where all selectable values cover "all caches" this is identical to the behaviour when all choices are selected.

The behavior is fine in general, but is not fitting to the cache type filter. Why can't we just make a exception for cache type? 

BTW, we already have such an exception: If you add a status filter and deselect "enabled", "disabled" and "archivd" no caches are shown instead of all, which is at least from user perspective the same thing.

**Comment by fm-sys** [comment]
> But using the filtering system to get rid of displayed caches is a really bad idea - because e.g. in live map c:geo will still issue requests to the servers

 Does this mean, that GC and OC caches are still requested, even if only Lab Caches are selected as cache type? If so, this should be improved anyway independent of this issue...

**Comment by eddiemuc** [comment]
I am 100% with @moving-bits on this. Having some sort of "hide all" option in map quick settings (which also ends all server calls) would be very valuable to have

**Comment by moving-bits** [comment]
Here's a proposal for how such a "hide all" checkbox could be integrated into map quick settings:

- visual integration:
  - switch the whole first block from "show ..." to "hide ..." and adapt subheadings accordingly (as discussed earlier)
  - add an extra subheading "Hide" at the beginning
  - add a "all" checkbox in top-right position
  - if "hide all" is activated, all other options in the first block are disabled (and vv)
- functional integration:
  - an activated "hide all" would overrule all other filters, but without changing their settings
  - "hide all" is a transient setting valid only in current map session; its value does not get persisted
  - when activated, no cache and no waypoint is being shown on the map; one could discuss whether tracks and routes also get hidden

|"hide all" disabled|"hide all" enabled|
|---|---|
|![image](https://user-images.githubusercontent.com/3754370/129480896-8014e140-23f6-47f7-83d3-93f219fb16f9.png)|![image](https://user-images.githubusercontent.com/3754370/129480899-74da73b3-eaf9-452c-bfe1-44dbe17311c7.png)|

This would also help with
- possible linebreak in "Show waypoints" subheading (depending on language)
- misunderstanding of map quick settings in context of filters (a "show xxx" still cannot display things which are filtered out by some filter rule, therefore "hide xxx" is easier to understand)

**Comment by moving-bits** [comment]
If this is only about getting a view of the whole map for a quick orientation, you can use compact icon mode to achieve this already: just switch it to `on`, and you'll see much more of the map.

(Still https://github.com/cgeo/cgeo/issues/10963#issuecomment-899055010 would give us some additional benefits, at the price of inverting the display logic once again. But for now the workaround with compact icon mode spares us code & behavior changes.)

**Comment by fm-sys** [comment]
Compact mode is at least better than the normal version, but it's still annoying if you try to find a specific place on the map in areas you do not know very well (while not being interested in any geocache at all). 

Imagine Google Maps would be that cluttered with all types of "special locations" (while I'm not saying Lidl & Edeka are really the most important place there, it's at least more cleaned up ;-)) 

||||
-|-|-
![Screenshot_20211106_190432_cgeo geocaching](https://user-images.githubusercontent.com/64581222/140619483-1bbf197b-6775-4cd8-8a78-10049da47f94.jpg)|![Screenshot_20211106_190420](https://user-images.githubusercontent.com/64581222/140619485-8616a1f1-8b44-4edc-9814-b3c3bbf5e8a8.jpg)|![Screenshot_20211106_191254](https://user-images.githubusercontent.com/64581222/140619678-6dc8cd5e-7347-4524-a329-675d11b078b4.jpg)

**Comment by fm-sys** [comment]
Furthermore it would be cool if we could hide all caches and all waypoints separately...

**Comment by moving-bits** [comment]
Reminder to myself (or whoever is going to implement #12928): Depending on how we implement this in `UnifiedMap` this could be as easy as enabling/disabling a single layer (for caches and/or another one for waypoints).
